### PR TITLE
cursor handling transferred from Rapid V1 to V2

### LINKED ISSUE #777: Investigation: Cursor handling from v1 -> v2 (CSS-based to Pixi-fied)
> As a 'first issue', here is our [getting started page](https://github.com/facebook/RapiD/wiki/New-Developer-Introduction).

in the process of rewriting the renderer, we no longer need to rely on CSS styling rules to manipulate the cursors- pixi can do that for us. 

Here a couple of examples of how to change a cursor in using Pixi: 
- https://pixijs.io/examples/#/events/custom-mouse-icon.js
- https://pixijs.io/examples/#/sprite/texture-swap.js

Code Points: 
[55_cursors.css](https://github.com/facebook/RapiD/blob/main/css/55_cursors.css) contains the .css code for these cursors. 

[context.js](https://github.com/facebook/RapiD/blob/main/modules/core/context.js#L358) contains the exact line where the mode sets the map container's .css class, which drives the cursor changes. 

[PixiTextures.js](https://github.com/facebook/RapiD/blob/main/modules/pixi/PixiTextures.js#L31) will need modification to accomplish the extra credit deliverable- you'll need to use the asset loader to load the cursor icons and make them addressable from the PIXI rendering subsystem.  You should be able to use  something like 'Load Patterns' code on [line 39](https://github.com/facebook/RapiD/blob/main/modules/pixi/PixiTextures.js#L39) to load all the cursors from the top-level `img` directory. 

Some of the classes you'll see are named like:
`.ideditor.mode-add-preset .main-map`
`.ideditor.mode-draw-line .main-map`

This means that the .css rule for the cursor will be applied in the condition that the `.main-map` container has the class `.mode-*`. These modes are shuffled / changed as the user clicks around the map and does things. 

More about modes:  You can think of a mode as a task that the user is accomplishing, like drawing a shape or moving a point around.  The new rapid has some slightly different mode names than the old one, so there may be some modes in the .css that don't exist anymore. So you will need to familiarize yourself with how rapid USED to work, and then figure out how RapiD *should* work in v2. 

For example- rapid v1 had modes for 'add area' and 'draw area'- now we just have a single mode for 'draw area'. Same with 'draw way' - we've collapsed the modes from 2 (add and draw way) to 1 (just draw way). 

Suggest that you get two copies of VSCode up and running so that you can compare and contrast your v2 and v1 copies.  More on this is covered in the last note of the [getting started page](https://github.com/facebook/RapiD/wiki/New-Developer-Introduction).

Now that you have two copies of rapid, you can run them both simultaneously and compare how they operate. 

### Doing the work

Okay, we want to ensure that we have the same cursor experience from 1.x to 2.x. So the very first item to complete is to run 1.x in the debug environment and understand all the ways the cursor can change. 

Tip: Do a global search for 'cursor' in the .css code and try to find all the places where it is switched. Then work backwards from the cursor names to .css rules, then figure out where those rules are changed. 

There are at least two big classes of code where this cursor change may occur:
- Mode Changes
- Streetview Photo Editor interactions

There may be others.  For now, focus on the mode changes, as these mode changes affect the cursor IN THE MAIN MAP SURFACE, which is now webGL-ified. Leave the streetview photo editor interactions alone, as that is an 'overlay' using regular .css content above the webGL map. 

### Deliverables: 
- [x] a list of all the code pointers and circumstances under which the main map cursor is changed in the old (v1) code. and a potential list of places where the code can be modified in the new code. 
- [ ] Extra Credit: A Pull Request to the main branch with the pixi-fied cursor transitions implemented, and a PR to the main branch containing / explaining the changes. 

### Impact of this work
We're trying to divest ourselves of as much .css as possible, so that there is less cognitive load on us for figuring out 'where' things happen in the renderer.  By splitting out the cursor code from .css, we can improve our dev experience by removing unused code (seriously, there's nothing worse than code that's sitting there undeleted!) and we can unlock better/more awesomer cursor experiences.

### PR DESCRIPTION
**Pull request by tannerwuster**
### Rapid V1 Cursor Handling add to Rapid V2
**This includes the following hover effects:**
- Lasso tool 
- Drawing a connect line 
- Connecting a vertex 
- Select Add
- Select Area 
- Select Line 
- Select Mapillary 
- Select Point 
- Select Remove
- Select Split 
- Select Vertex
_I believe this still needs another iteration that uses the Pixi textures to load the images_
closes #777

### CODE REVIEW

File: modules/modes/ModeBrowse.js [hunk 1]
```diff
@@ -64,5 +66,37 @@ export class ModeBrowse extends AbstractMode {
...
+    const graph = context.graph();
+    // Get the target and associated datum
+    const target = eventData.target;
+    const datum = target && target.data;
+    // Check if the datum is an entity in the graph
+    const entity = datum && graph.hasEntity(datum.id);
+    // Get the geometry of the entity, if it exists
+    const geom = entity && entity.geometry(graph);
+    // Change the cursor of the document body based on the geometry type
+    if (geom && geom === 'vertex') {
```
Comment (Bonkles) [inline modules/modes/ModeBrowse.js:88]: In general, when decision making in an if/else, try to limit things to a single assignment statement at the end (if possible). 

```
let cursorPath;
if (vertex) {
cursorPath = 'url(/img/cursor/cursor-select-vertex.png), auto';
} 
/* ... other clauses cut for brevity  ... */ 
else {
    cursorPath = 'url(/img/cursor/cursor-grab.png),auto';
}

document.body.style.cursor = cursorPath
}

```

File: modules/modes/ModeBrowse.js [hunk 2]
```diff
@@ -64,5 +66,37 @@ export class ModeBrowse extends AbstractMode {
...
+      document.body.style.cursor = 'url(/img/cursor/cursor-select-vertex.png),auto';
+    } else if (geom && geom === 'line') {
+      document.body.style.cursor = 'url(/img/cursor/cursor-select-line.png),auto';
+    } else if (geom && geom === 'area') {
+      document.body.style.cursor = 'url(/img/cursor/cursor-select-area.png),auto';
+    } else if (geom && geom === 'point') {
+      document.body.style.cursor = 'url(/img/cursor/cursor-select-point.png),auto';
+    } else {
+      // If there is no entity or the entity's geometry is unknown, use the grab cursor
+      document.body.style.cursor = 'url(/img/cursor/cursor-grab.png),auto';
```
Comment (Bonkles) [inline modules/modes/ModeBrowse.js:98]: But in this particular case, we can shorten the code to be even more brief: 

```
if (geom && (geom === 'line' || geom === 'area' || geom === 'point'))  {
 document.body.style.cursor = `url(/img/cursor/cursor-select-${geom}.png),auto`;
 }  else {
 document.body.style.cursor = 'url(/img/cursor/cursor-grab.png),auto';
 }
```

File: modules/modes/ModeDrawArea.js [hunk 3]
```diff
@@ -558,4 +559,30 @@ if (choice && choice.distance < SNAP_DIST) {
...
+    // Get the target and associated datum
+    const target = eventData.target;
+    const datum = target && target.data;
+    // Check if the datum is an entity in the graph
+    const entity = datum && graph.hasEntity(datum.id);
+    // Get the geometry of the entity, if it exists
+    const geom = entity && entity.geometry(graph);
+    // Change the cursor of the document body based on the geometry type
+    if (geom && geom === 'vertex') {
+      document.body.style.cursor = 'url(/img/cursor/cursor-draw-connect-vertex.png),auto';
```
Comment (Bonkles) [inline modules/modes/ModeDrawArea.js:580]: I built and ran this branch, and I noticed that the draw cursors are indeed showing up but they're not offset correctly. All the crosshair-style cursors (the ones that begin with 'cursor-draw') will need to be centered directly on the mouse coordinate, unlike the pointer-type that need to be drawn underneath/to the right.

**Review by Bonkles — CHANGES_REQUESTED** [review]
This is a pretty gnarly problem with lots of use cases! Though, I am impressed you've gotten this far with no help. Solid work! You should be very happy. 


A few things here & there, see my inline comments. Some are general code structure/suggestions, some are actual issues i saw in the behavior. 

We will need to take another look at the cursor transitions as you click on a thing, then hover on that thing, then hover off that thing. It doesn't look like the cursor transitions are quite correct in all cases.

**Comment by bhousel** [comment]
Going to close this, the code has changed a bunch so it's gone stale.

**Pull request closed without merging**